### ISSUE TITLE
Span<T> support

### ISSUE THREAD

**Issue by lanwin** [AngleSharp/AngleSharp#619]
https://channel9.msdn.com/Events/Connect/2017/T125

I think supporting Span<T> could be very helpful for a library like which does o lot with strings :)

**Comment by FlorianRappl** [comment]
Yes, but besides the language feature `Span<T>` is also a framework feature. Needs the SpanExtensions, which would naturally exclude, e.g., .NET 4 and SL. Since support for these will potentially be discontinued anyway, a rewrite / use may be feasible.

Should be first assessed in a POC, as I am sure memory consumption would be nice, but I am not yet convinced performance wise (most of the time we are on a char, i.e., value type, level, but with slices we have to de-reference every time, which may may be up to 2x slower; needs to be investigated which effect will kick in first).

**Comment by FlorianRappl** [comment]
Quick update: Last year I looked into this. For the HTML parsing it is not possible to gain any win here (due to similar reasons as I'll point out in a second), so I tried it on the URL parser first. This one operates on a given string and just tries to move some pointer forward (consuming the characters 1 by 1 with rules coming from an internal state machine).

Long story short: Since many comparisons had to be case insensitive or in a regular expression like manner (one of these must fit), i.e., not requiring an exact match, the implementation required many "non-standard" things from `Span<T>`. Quite often a substring had to be gathered or manual (i.e., using a loop) comparisons had to be done. This was all quite inefficient (at least it felt that way). In the end the code was more complicated and performed even worse (round about the same for small URLs, but getting worse with increased length). As a result I dropped this.

HTML will have the same problems as many things need either insensitive comparisons or very special rules. Furthermore, using `Span<T>` we cannot get rid of the internal `StringBuilder` buffer, which is needed to insert things (e.g., coming from the `document.write` API) into the running context. Also since we may have encoding problems (or coming from a real byte-stream) the buffer still makes sense performance-wise (here we already have UTF-16, i.e., ".NET native" characters).

I'll close this for now until I see a real chance of applying it somewhere (in AngleSharp).
